libs/adapters/src/rabbitmq/RabbitMQAdapter.ts

```diff
@@ -16,30 +17,63 @@ import { Message } from 'amqplib';
 import { AckOrNack, default as Rascal } from 'rascal';
 import { fileURLToPath } from 'url';
 
-const defaultRetryStrategy: RetryStrategyFn = (
-  err: Error | undefined,
-  topic: BrokerSubscriptions,
-  content: any,
-  ackOrNackFn: AckOrNack,
-  log: ILogger,
-) => {
-  if (err instanceof InvalidInput) {
-    log.error(`Invalid event`, err, {
-      topic,
-      message: content,
-    });
-    ackOrNackFn(err, { strategy: 'nack' });
-  } else {
-    log.error(`Failed to process event`, err, {
+/**
+ * Build a retry strategy function based on custom retry strategies map.
+ *
+ * @param {Function} customRetryStrategiesMap - A function which maps errors to retry strategies. The function should
+ * return `true` if an error was successfully mapped to a strategy and `false` otherwise.
+ * @param defaultDefer
+ * @param defaultAttempts
+ * @returns {RetryStrategyFn} The built retry strategy function.
+ */
+export function buildRetryStrategy(
+  customRetryStrategiesMap?: (...args: Parameters<RetryStrategyFn>) => boolean,
+  defaultDefer: number = 2000,
+  defaultAttempts: number = 3,
+): RetryStrategyFn {
+  return function (
+    err: Error | InvalidInput | CustomRetryStrategyError,
+    topic: BrokerSubscriptions,
+    content: any,
+    ackOrNackFn: AckOrNack,
+    log: ILogger,
+  ) {
+    const logContext = {
       topic,
       message: content,
-    });
-    ackOrNackFn(err, [
-      { strategy: 'republish', defer: 2000, attempts: 3 },
-      { strategy: 'nack' },
-    ]);
-  }
-};
+    };
+
+    if (err instanceof InvalidInput) {
+      log.error(`Invalid event`, err, logContext);
+      ackOrNackFn(err, { strategy: 'nack' });
+      return;
+    } else if (err instanceof CustomRetryStrategyError) {
+      log.error(err.message, err, logContext);
+      ackOrNackFn(err, err.recoveryStrategy);
+      return;
+    }
+
+    let res = false;
+    if (customRetryStrategiesMap) {
+      res = customRetryStrategiesMap(err, topic, content, ackOrNackFn, log);
+    }
+
+    if (!res) {
```

Comment:
Why not making strategies map a mandatory argument? seems like we will always need a strategy (with a sensible default)... then this like would look like:
```
if( ! retryStrategies(err, topic, content, ...)) {
...
}
```

Reply:
Actually, the opposite seems to be true. We almost always don't need a strategy since the default retry X times in Y intervals default covers most cases. For the contest projection handlers for example we only need to modify the default retry strategy to increase X and Y and don't need to add an additional strategy for specific errors.

Reply:
ok.. I was thinking about an explicit default strategy to make the strategy argument required

Reply:
I mean an explicit default argument like 

```
function build(strategy = defaultStrategy) {
...
}
```

Reply:
Gonna go ahead and resolve/merge this so we can move forward to see if this will fix the race condition.